### PR TITLE
Add a tiled big image carousel module.

### PR DESCRIPTION
**Pull request by plural**

### CODE REVIEW

File: demo_modules/big_image_carousel/big_image_carousel.js [hunk 1]
```diff
@@ -0,0 +1,199 @@
+/* Copyright 2015 Google Inc. All Rights Reserved.
```
Comment (applmak) [inline demo_modules/big_image_carousel/big_image_carousel.js]: TODO, bump this year.

File: demo_modules/big_image_carousel/big_image_carousel.js [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+limitations under the License.
+==============================================================================*/
+
+const ModuleInterface = require('lib/module_interface');
+const debug = require('debug');
+const register = require('register');
+const wallGeometry = require('wallGeometry');
+
+class BigImageCarouselServer extends ModuleInterface.Server {
+  constructor(config) {
```
Comment (applmak) [inline demo_modules/big_image_carousel/big_image_carousel.js]: Rm this constructor

File: demo_modules/big_image_carousel/big_image_carousel.js [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    this.surface = surface;
+    this.tiles = null; 
+
+    this.image_list_config = extraArgs;
+  }
+
+  preload() {
+    var p5 = this.p5;
+
+    console.log('preload');
```
Comment (applmak) [inline demo_modules/big_image_carousel/big_image_carousel.js]: Prefer debug to console.log

File: demo_modules/big_image_carousel/big_image_carousel.js [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+    this.y_index = this.surface.virtualRect.y / this.surface.virtualRect.h;
```
Comment (applmak) [inline demo_modules/big_image_carousel/big_image_carousel.js]: This isn't _necessarily_ valid. Consider this.surface.virtualOffset.y

File: demo_modules/big_image_carousel/big_image_carousel.js [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+    console.log('screen with y_index ' + this.y_index);
+
+    const asset = require('client/asset/asset');
+
+    this.tiles = new Array();
```
Comment (applmak) [inline demo_modules/big_image_carousel/big_image_carousel.js]: Prefer functional code:

this.tiles = Object.keys(this.image_list_config).map(path => {
  ...
})

File: demo_modules/big_image_carousel/big_image_carousel.js [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+          this.tiles.push(p5.loadImage(tilePath));
+        }
+      }
+    }
+  }
+
+  setup() {
+    var p5 = this.p5;
+
+    console.log('setup()');
```
Comment (applmak) [inline demo_modules/big_image_carousel/big_image_carousel.js]: rm

File: demo_modules/big_image_carousel/big_image_carousel.js [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    this.sum_tile_width = 0;
+    for (let x = 0; x < this.tiles.length; x++) {
+      this.sum_tile_width += this.tiles[x].width;
+    }
+    console.log('num tiles is ' + this.tiles.length);
+    console.log('sum_tile_width is ' + this.sum_tile_width);
+    if (this.sum_tile_width < this.surface.wallRect.w) {
+      let initial_sum_tile_width = this.sum_tile_width;
+      while (this.sum_tile_width < this.surface.wallRect.w) {
+        this.tiles = this.tiles.concat(this.tiles);
```
Comment (applmak) [inline demo_modules/big_image_carousel/big_image_carousel.js]: Huh? What does this do?

File: demo_modules/big_image_carousel/big_image_carousel.js [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+    }
+
+    if (this.y_index == null) {
+      console.log('null y_index');
+      return;
+    }
+
+    p5.background(120);
+
+    let x_zero = (((t - this.surface.startTime) / 9) % (2*this.sum_tile_width)) - this.sum_tile_width;
```
Comment (applmak) [inline demo_modules/big_image_carousel/big_image_carousel.js]: Doc this expression

File: demo_modules/big_image_carousel/big_image_carousel.js [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+    // for each iteration:
+    // - translate x zero
```
Comment (applmak) [inline demo_modules/big_image_carousel/big_image_carousel.js]: x_zero

File: demo_modules/big_image_carousel/big_image_carousel.js (hunk 8 above)
Comment (applmak) [inline demo_modules/big_image_carousel/big_image_carousel.js]: x_0 is better than x_zero

File: demo_modules/big_image_carousel/big_image_carousel.js [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    // if sum of all image widths > wall width, identify the starting 
+    let x_offset = x_zero;
+    // walk backwards through the list to draw the tail end until we have no more of it.
+//    for (let x = 0; x
+    // Draw the "positive" images.
+    for (let x = 0; x < this.tiles.length; x++) {
+      let image = this.tiles[x];
+      let right_x = x_offset + image.width;
+        // Do we need to draw this image?
+        if ((x_offset >= this.surface.virtualRect.x &&
```
Comment (applmak) [inline demo_modules/big_image_carousel/big_image_carousel.js]: Prefer filter than forEach to foreach than filter

File: demo_modules/big_image_carousel/big_image_carousel.js [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+    if (x_zero < 0) {
+      // need to fill in images on the right.
+      while (x_offset < this.surface.wallRect.w) {
+        for (let x = 0; x < this.tiles.length; x++) {
+          let image = this.tiles[x];
+          let right_x = x_offset + image.width;
+            // Do we need to draw this image?
+            if ((x_offset >= this.surface.virtualRect.x &&
+                x_offset <= this.surface.virtualRect.x + this.surface.virtualRect.w)
```
Comment (applmak) [inline demo_modules/big_image_carousel/big_image_carousel.js]: Save virtualRect.x + w as a local const var

File: demo_modules/big_image_carousel/big_image_carousel.js [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+          if ((x_offset >= this.surface.virtualRect.x &&
+              x_offset <= this.surface.virtualRect.x + this.surface.virtualRect.w)
+              ||
+              (right_x >= this.surface.virtualRect.x &&
+               right_x < this.surface.virtualRect.x + this.surface.virtualRect.w)) {
+            p5.image(image, x_offset, this.surface.virtualRect.y);
+          }
+      }
+    }
+      p5.text("X Zero is " + x_zero, this.surface.virtualRect.x + 50, this.surface.virtualRect.y + 100);
```
Comment (applmak) [inline demo_modules/big_image_carousel/big_image_carousel.js]: Debug?

File: demo_modules/big_image_carousel/big_image_carousel.js [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+  }
+}
+
+class BigImageCarouselClient extends ModuleInterface.Client {
+  constructor(config) {
+    super();
+    this.config = config;
+    debug('BigImageCarousel Client!', config);
```
Comment (applmak) [inline demo_modules/big_image_carousel/big_image_carousel.js]: rm log spam

File: demo_modules/big_image_carousel/big_image_carousel.js (hunk 1 above)
Comment (plural) [inline demo_modules/big_image_carousel/big_image_carousel.js]: Done

File: demo_modules/big_image_carousel/big_image_carousel.js (hunk 2 above)
Comment (plural) [inline demo_modules/big_image_carousel/big_image_carousel.js]: done

File: demo_modules/big_image_carousel/big_image_carousel.js (hunk 3 above)
Comment (plural) [inline demo_modules/big_image_carousel/big_image_carousel.js]: done throughout.

File: demo_modules/big_image_carousel/big_image_carousel.js (hunk 4 above)
Comment (plural) [inline demo_modules/big_image_carousel/big_image_carousel.js]: punting on this for now.

File: demo_modules/big_image_carousel/big_image_carousel.js (hunk 5 above)
Comment (plural) [inline demo_modules/big_image_carousel/big_image_carousel.js]: added a TODO.

File: demo_modules/big_image_carousel/big_image_carousel.js (hunk 6 above)
Comment (plural) [inline demo_modules/big_image_carousel/big_image_carousel.js]: done

File: demo_modules/big_image_carousel/big_image_carousel.js (hunk 7 above)
Comment (plural) [inline demo_modules/big_image_carousel/big_image_carousel.js]: it basically extends the tile list to have enough tiles to be wider than the screen.

File: demo_modules/big_image_carousel/big_image_carousel.js (hunk 8 above)
Comment (plural) [inline demo_modules/big_image_carousel/big_image_carousel.js]: done and done.

File: demo_modules/big_image_carousel/big_image_carousel.js (hunk 9 above)
Comment (plural) [inline demo_modules/big_image_carousel/big_image_carousel.js]: done

File: demo_modules/big_image_carousel/big_image_carousel.js (hunk 11 above)
Comment (plural) [inline demo_modules/big_image_carousel/big_image_carousel.js]: done

**Comment by plural** [comment]
punted on a couple of things.

File: demo_modules/big_image_carousel/big_image_carousel.js (hunk 12 above)
Comment (plural) [inline demo_modules/big_image_carousel/big_image_carousel.js]: removed

File: demo_modules/big_image_carousel/big_image_carousel.js (hunk 13 above)
Comment (plural) [inline demo_modules/big_image_carousel/big_image_carousel.js]: removed

**Pull request merged**